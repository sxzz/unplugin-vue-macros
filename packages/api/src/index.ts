export { MagicString } from '@vue-macros/common'

// WIP
// export * from './vue'
export * from './ts'
export * from './utils'
