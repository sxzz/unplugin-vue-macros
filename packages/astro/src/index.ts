import { type AstroIntegration } from 'astro'
import { type Options } from 'unplugin-vue-macros'
import VitePlugin from 'unplugin-vue-macros/vite'

const name = 'unplugin-vue-macros'

export default function VueMacrosAstroIntegration(
  userOptions: Options = {}
): AstroIntegration {
  return {
    name,
    hooks: {
      'astro:config:setup': ({ updateConfig }) => {
        updateConfig({
          vitePlugin: [VitePlugin(userOptions)],
        })
      },
    },
  }
}
