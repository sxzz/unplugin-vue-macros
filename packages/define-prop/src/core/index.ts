import {
  HELPER_PREFIX,
  DEFINE_PROP,
  DEFINE_PROPS,
  MagicString,
  getTransformResult,
  isCallOf,
  parseSFC,
  walkAST,
  importHelperFn,
} from '@vue-macros/common'
import type { Node } from '@babel/types'


function mountProps(props: string[][]) {
  const isAllWithoutOptions = props.every(([, options]) => !options)

  if (isAllWithoutOptions) {
    return `[${props.map(([name]) => `'${name}'`).join(',')}]` 
  }

  return `{
    ${props.map(([name, options]) => `${name}: ${options || `{}`}`).join(', ')}
  }`
}

export const propsVariableName = `${HELPER_PREFIX}_root_props`

export function transformDefineProp(code: string, id: string) {
  if (!code.includes(DEFINE_PROP)) return

  const { scriptSetup, getSetupAst } = parseSFC(code, id)
  if (!scriptSetup) return

  const offset = scriptSetup.loc.start.offset
  const s = new MagicString(code)
  const setupAst = getSetupAst()!

  const props: string[][] = []

  
  
  walkAST<Node>(setupAst, {
    enter(node) {
      if (isCallOf(node, DEFINE_PROP)) {

        let options = undefined

        if (node.arguments[1]) {
          options = code.slice(node.arguments[1].start! + offset, node.arguments[1].end! + offset)
        }


        props.push([node.arguments[0].value, options])

        
        s.overwriteNode(
          node,
          // add space for fixing mapping
          `${HELPER_PREFIX}computed(() => ${propsVariableName}.${node.arguments[0].value})`,
          { offset }
        )
      }
    },
  })

  if (props.length) {
    importHelperFn(s, offset, 'computed', 'vue')

    s.prependLeft(offset!, `\n const ${propsVariableName} = defineProps(${mountProps(props)})\n`)
  }

  return getTransformResult(s, id)
}
