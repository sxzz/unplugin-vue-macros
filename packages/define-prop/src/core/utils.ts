import type * as t from '@babel/types'
import type { MagicStringAST } from '@vue-macros/common'

export type Impl = (ctx: {
  s: MagicStringAST
  offset: number
  resolveTSType: (type: t.TSType) => Promise<string[] | undefined>
}) => {
  walkCall: (node: t.CallExpression, parent?: t.Node | null) => string
  genRuntimeProps: (isProduction: boolean) => Promise<string | undefined>
}

export function stringifyArray(strs: string[]): string {
  return `[${strs.map((s) => JSON.stringify(s)).join(', ')}]`
}
