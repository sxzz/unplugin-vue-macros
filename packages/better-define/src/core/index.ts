import { MagicString, getTransformResult, parseSFC } from '@vue-macros/common'
import { analyzeSFC } from '@vue-macros/api'
import type { TSProps } from '@vue-macros/api'
import type {} from '@babel/types'

export const transformBetterDefine = async (code: string, id: string) => {
  const s = new MagicString(code)
  const sfc = parseSFC(code, id)
  if (sfc.script || !sfc.scriptSetup) return

  const offset = sfc.scriptSetup.loc.start.offset
  const result = await analyzeSFC(s, sfc)
  if (result.props) {
    processProps(result.props)
  }

  return getTransformResult(s, id)

  async function processProps(props: TSProps) {
    const runtimeDefs = await props.getRuntimeDefinitions()

    // TODO prod, default
    const runtimeDecls = `{\n  ${Object.entries(runtimeDefs)
      .map(
        ([key, { type, required }]) =>
          `${key}: { type: ${toRuntimeTypeString(
            type
          )}, required: ${required} }`
      )
      .join(',\n  ')}\n}`

    s.overwriteNode(props.callExpressionAst, `defineProps(${runtimeDecls})`, {
      offset,
    })
  }

  function toRuntimeTypeString(types: string[]) {
    return types.length > 1 ? `[${types.join(', ')}]` : types[0]
  }
}
