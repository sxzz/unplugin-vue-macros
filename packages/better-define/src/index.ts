import process from 'node:process'
import { RollupResolve, setResolveTSFileIdImpl } from '@vue-macros/api'
import {
  createFilter,
  detectVueVersion,
  FilterFileType,
  getFilterPattern,
  type BaseOptions,
  type MarkRequired,
} from '@vue-macros/common'
import {
  createUnplugin,
  type UnpluginContextMeta,
  type UnpluginInstance,
} from 'unplugin'
import { generatePluginName } from '#macros' with { type: 'macro' }
import { transformBetterDefine } from './core'
import type { PluginContext } from 'rollup'

export type Options = BaseOptions
export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'isProduction'
>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
    framework,
  )
  return {
    include,
    isProduction: process.env.NODE_ENV === 'production',
    ...options,
    version,
  }
}

const name = generatePluginName()

const plugin: UnpluginInstance<Options | undefined, false> = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    const { resolve, handleHotUpdate } = RollupResolve()

    return {
      name,
      enforce: 'pre',

      buildStart() {
        if (framework === 'rollup' || framework === 'vite') {
          setResolveTSFileIdImpl(resolve(this as PluginContext))
        }
      },

      transformInclude: filter,
      async transform(code, id) {
        try {
          return await transformBetterDefine(code, id, options.isProduction)
        } catch (error: unknown) {
          this.warn(`${name} ${error}`)
          console.warn(error)
        }
      },

      vite: {
        configResolved(config) {
          options.isProduction ??= config.isProduction
        },

        handleHotUpdate,
      },
    }
  },
)
export default plugin
