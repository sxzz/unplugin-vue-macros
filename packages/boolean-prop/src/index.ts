import { getVuePluginApi, type VuePluginApi } from '@vue-macros/common'
import { generatePluginName } from '#macros' with { type: 'macro' }
import { transformBooleanProp, type Options } from './core/index'
import type { Plugin } from 'vite'

// legacy export
export * from './api'

const name = generatePluginName()

function rollup(options: Options = {}): Plugin {
  let api: VuePluginApi | null | undefined

  return {
    name,
    configResolved(config) {
      try {
        api = getVuePluginApi(config.plugins)
      } catch {}
    },
    buildStart(rollupOpts) {
      if (api === undefined)
        try {
          api = getVuePluginApi(rollupOpts.plugins)
        } catch (error: any) {
          this.warn(error)
          return
        }

      if (!api) return

      api.options.template ||= {}
      api.options.template.compilerOptions ||= {}
      api.options.template.compilerOptions.nodeTransforms ||= []

      api.options.template.compilerOptions.nodeTransforms.push(
        transformBooleanProp(options),
      )
    },
  }
}

const plugin: {
  rollup: typeof rollup
  vite: typeof rollup
} = {
  rollup,
  vite: rollup,
}
export default plugin
