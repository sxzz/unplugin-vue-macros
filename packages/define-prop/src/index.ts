import process from 'node:process'
import { type UnpluginContextMeta, createUnplugin } from 'unplugin'
import {
  type BaseOptions,
  FilterFileType,
  type MarkRequired,
  createFilter,
  detectVueVersion,
  getFilterPattern,
  normalizePath,
} from '@vue-macros/common'
import { RollupResolve, setResolveTSFileIdImpl } from '@vue-macros/api'
import { generatePluginName } from '#macros' assert { type: 'macro' }
import { type Edition, transformDefineProp } from './core'
import { helperCode, helperId } from './core/helper'
import type { PluginContext } from 'rollup'

export interface Options extends BaseOptions {
  isProduction?: boolean
  edition?: Edition
}

export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'isProduction' | 'edition'
>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
    framework,
  )
  return {
    include,
    isProduction: process.env.NODE_ENV === 'production',
    edition: 'kevinEdition',
    ...options,
    version,
  }
}

const name = generatePluginName()

export default createUnplugin<Options | undefined, false>(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)
    const { resolve, handleHotUpdate } = RollupResolve()

    return {
      name,
      enforce: 'pre',

      buildStart() {
        if (framework === 'rollup' || framework === 'vite') {
          setResolveTSFileIdImpl(resolve(this as PluginContext))
        }
      },

      resolveId(id) {
        if (id === normalizePath(helperId)) return id
      },

      loadInclude(id) {
        return normalizePath(id) === helperId
      },

      load(id) {
        if (normalizePath(id) === helperId) return helperCode
      },

      transformInclude: filter,
      transform(code, id) {
        return transformDefineProp(
          code,
          id,
          options.edition,
          options.isProduction,
        )
      },

      vite: {
        configResolved(config) {
          options.isProduction = config.isProduction
        },

        handleHotUpdate,
      },
    }
  },
)
