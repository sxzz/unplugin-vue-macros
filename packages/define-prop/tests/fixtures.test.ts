import { testFixtures } from '@vue-macros/test-utils'
import { describe } from 'vitest'
import { transformDefineProp } from '../src/core'

describe('fixtures', async () => {
  await testFixtures(
    import.meta.glob('./fixtures/detalhed.{vue,js,ts}', {
      eager: true,
      as: 'raw',
    }),
    (args, id, code) => transformDefineProp(code, id)?.code
  )
})
