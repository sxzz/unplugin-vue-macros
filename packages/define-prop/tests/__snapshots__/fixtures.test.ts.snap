// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixtures > ./fixtures/detalhed.vue 1`] = `
"<script setup lang=\\"ts\\">
 const __MACROS__root_props = defineProps({
    foo: {
    type: String,
    required: true,
    default: 'foo'
}, bar: {
    type: [Number, Array],
    required: true,
    default: () => [1, 2, 3]
}
  })

import { computed as __MACROS_computed } from 'vue';
import { expectTypeOf } from 'expect-type'
import type { ComputedRef } from 'vue'

const foo = __MACROS_computed(() => __MACROS__root_props.foo)

const bar = __MACROS_computed(() => __MACROS__root_props.bar)

expectTypeOf(foo).toEqualTypeOf<ComputedRef<string>>()
expectTypeOf(bar).toEqualTypeOf<ComputedRef<number[]>>()

</script>

<template>
  <div></div>
</template>
"
`;
